Mage.Sets/src/mage/cards/r/RockHydra.java

```diff
@@ -0,0 +1,131 @@
+import mage.abilities.effects.common.PreventDamageToSourceEffect;
+import mage.abilities.effects.common.counter.AddCountersSourceEffect;
+import mage.cards.CardImpl;
+import mage.cards.CardSetInfo;
+import mage.constants.CardType;
+import mage.constants.Duration;
+import mage.constants.PhaseStep;
+import mage.constants.Zone;
+import mage.counters.CounterType;
+import mage.game.Game;
+import mage.game.events.GameEvent;
+import mage.game.permanent.Permanent;
+
+/**
+ *
+ * @author anonymous
+ */
+public class RockHydra extends CardImpl {
+
+    public RockHydra(UUID ownerId, CardSetInfo setInfo) {
+        super(ownerId, setInfo, new CardType[]{CardType.CREATURE}, "{X}{R}{R}");
+        
+        this.subtype.add("Hydra");
+        this.power = new MageInt(0);
+        this.toughness = new MageInt(0);
+        
+        // Rock Hydra enters the battlefield with X +1/+1 counters on it.
+        this.addAbility(new EntersBattlefieldAbility(new EntersBattlefieldWithXCountersEffect(CounterType.P1P1.createInstance())));
+        // For each 1 damage that would be dealt to Rock Hydra, if it has a +1/+1 counter on it, remove a +1/+1 counter from it and prevent that 1 damage.
+        this.addAbility(new SimpleStaticAbility(Zone.BATTLEFIELD, new RockHydraEffect()));
+        // {R}: Prevent the next 1 damage that would be dealt to Rock Hydra this turn.
+        this.addAbility(new SimpleActivatedAbility(Zone.BATTLEFIELD, new PreventDamageToSourceEffect(Duration.EndOfTurn, 1), new ManaCostsImpl("{R}")));
+        // {R}{R}{R}: Put a +1/+1 counter on Rock Hydra. Activate this ability only during your upkeep.
+        this.addAbility(new ConditionalActivatedAbility(Zone.BATTLEFIELD, new AddCountersSourceEffect(CounterType.P1P1.createInstance(1)), new ManaCostsImpl("{R}{R}{R}"), new IsStepCondition(PhaseStep.UPKEEP), null));
+    }
+
+    public RockHydra(final RockHydra card) {
+        super(card);
+    }
+
+    @Override
+    public RockHydra copy() {
+        return new RockHydra(this);
+    }
+    
+    class RockHydraEffect extends PreventionEffectImpl {
+
+        public RockHydraEffect() {
+            super(Duration.WhileOnBattlefield, Integer.MAX_VALUE, false, false);
+            staticText = "For each 1 damage that would be dealt to Rock Hydra, if it has a +1/+1 counter on it, remove a +1/+1 counter from it and prevent that 1 damage.";
+        }
+
+        public RockHydraEffect(final RockHydraEffect effect) {
+            super(effect);
+        }
+
+        @Override
+        public RockHydraEffect copy() {
+            return new RockHydraEffect(this);
+        }
+
+        @Override
+        public boolean apply(Game game, Ability source) {
+            return true;
+        }
+
+        @Override
+        public boolean replaceEvent(GameEvent event, Ability source, Game game) {
+            PreventionEffectData preventionEffectData = preventDamageAction(event, source, game);
+            if (preventionEffectData.getPreventedDamage() > 0) {

```

Comment:
The prevented damage will always be >=0, so you should drop the if statement completely. You can also get the damage that is tried to be applied with event.getAmount(). That will let you remove a number of counters equal to the damage assigned, not only prevented.

If a previous replacement effect replaced the effect entirely, no other replacement effects are fired. If it replaces it in part, it will change the event.getAmount() to the new amount.

The AI can be finicky sometimes. You might have better luck with running two instances of the client and making a player vs. player game.

PS: You might be interested in joining out Gitter room (https://gitter.im/magefree/mage). Most developers hang in there and can help you if you have questions.